File: ttviz/src/ttviz.cpp
#include "ttviz.hpp"
#include <audioviz/fx/Blur.hpp>
#include <audioviz/fx/Mult.hpp>
#include <audioviz/util.hpp>
#include <stdexcept>

ttviz::ttviz(const sf::Vector2u size, audioviz::Media &media, const int fft_size, const int antialiasing)
	: Base{size},
	  media{media},
	  fa{fft_size},
	  sa{sample_rate_hz, fft_size}
{
	// Check for stereo audio
	if (media.audio_channels() != 2)
		throw std::runtime_error{
			"ttviz requires stereo (2-channel) audio; got " + std::to_string(media.audio_channels()) + " channel(s)"};

	// background
	if (media.attached_pic() || media.has_video_stream())
	{
		auto &bg_layer = emplace_layer<audioviz::PostProcessLayer>("bg", size);

		if (media.has_video_stream())
			bg_txr = sf::Texture{size};
		else if (media.attached_pic())
			set_background(*media.attached_pic());

		bg_layer.add_draw({bg_spr});
	}

	// particles
	auto &particles_layer = emplace_layer<audioviz::PostProcessLayer>("particles", size);
	particles_layer.add_draw({ps});
	particles_layer.set_fx_cb(
		[](auto &orig_rt, auto &fx_rt, auto &target)
		{
			target.draw(fx_rt.sprite(), sf::BlendAdd);
			target.draw(orig_rt.sprite(), sf::BlendAdd);
		});

	// spectrum
	ss.set_left_backwards(true);
	auto &spectrum_layer = emplace_layer<audioviz::PostProcessLayer>("spectrum", size);
	spectrum_layer.add_draw({ss});
	spectrum_layer.set_fx_cb(
		[this](auto &orig_rt, auto &fx_rt, auto &target)
		{
			if (spectrum_bm)
			{
				target.draw(fx_rt.sprite(), *spectrum_bm);
				target.draw(orig_rt.sprite(), *spectrum_bm);
			}
			else
			{
				target.draw(fx_rt.sprite(), audioviz::util::GreatAmazingBlendMode);
				target.draw(orig_rt.sprite());
			}
		});

	// metadata
	title_text.setStyle(sf::Text::Bold | sf::Text::Italic);
	title_text.setCharacterSize(24);
	title_text.setFillColor({255, 255, 255, 150});

	artist_text.setStyle(sf::Text::Italic);
	artist_text.setCharacterSize(24);
	artist_text.setFillColor({255, 255, 255, 150});

	metadata.use_metadata(media);
	if (media.attached_pic())
		metadata.set_album_cover(*media.attached_pic(), {150, 150});
	metadata.set_position({30, 30});

	emplace_layer<audioviz::Layer>("metadata").add_draw({metadata});
}

void ttviz::update(std::span<const float> audio_buffer)
{
	capture_time("fft", sa.execute_fft(fa, audio_buffer));
	capture_time("spectrum_update", ss.update(fa, sa, bp, ip));
	capture_time("ps_update", ps.update(sa));

	if (media.has_video_stream())
	{
		const int frames_to_wait = framerate / media.video_framerate();
		if (vfcount < frames_to_wait)
			++vfcount;
		else
		{
			if (media.read_video_frame(bg_txr))
				bg_spr = bg_txr;
			vfcount = 1;
		}
	}

	color.increment_wheel_time();
}

void ttviz::add_default_effects()
{
	if (const auto bg = get_layer_as<audioviz::PostProcessLayer>("bg"))
	{
		bg->add_effect(
			media.has_video_stream() ? new audioviz::fx::Blur{2.5, 2.5, 5} : new audioviz::fx::Blur{7.5, 7.5, 15});

		if (!media.has_video_stream())
			bg->add_effect(new audioviz::fx::Mult{0.75});

		if (media.attached_pic())
			set_background(*media.attached_pic());
	}

	get_layer_as<audioviz::PostProcessLayer>("particles")->add_effect(new audioviz::fx::Blur{1, 1, 10});
	get_layer_as<audioviz::PostProcessLayer>("spectrum")->add_effect(new audioviz::fx::Blur{3, 3, 10});
}

void ttviz::set_album_cover(const std::string &image_path, const sf::Vector2f size)
{
	metadata.set_album_cover(sf::Texture{image_path}, size);
}

void ttviz::set_background(const sf::Texture &txr)
{
	bg_spr = bg_txr = txr;
	bg_spr.capture_centered_square_view();
	bg_spr.fill_screen(size);
}

void ttviz::set_spectrum_blendmode(const sf::BlendMode &bm)
{
	spectrum_bm = bm;
}

void ttviz::set_fft_size(int fft_size)
{
	fa.set_fft_size(fft_size);
	sa.set_fft_size(fft_size);
}

File: test/ttviz.lua
---@diagnostic disable: lowercase-global, undefined-global
package.cpath = package.cpath .. ';build/luaviz/?.so'
luaviz = require('luaviz')

if not arg[1] then
	print('media file required')
	os.exit(1)
end

-- video window size
SIZE = { 1280, 720 }

-- fft window size
FFT_SIZE = 3000

-- frequency analyzer (fft) logic
fa = luaviz.FrequencyAnalyzer.new(FFT_SIZE)
-- fa:set_accum_method('sum')

-- handles fft of stereo audio for you
sa = luaviz.StereoAnalyzer.new()

-- self-explanatory
cs = luaviz.ColorSettings.new()

-- the star of the show
bss = luaviz.BarStereoSpectrum.new({ {}, SIZE }, cs)
-- bss:set_bar_width(1)
-- bss:set_bar_spacing(0)

-- particles!!!!!!!!!!!
cps = luaviz.CircleParticleSystem.new({ {}, SIZE }, 50)

-- where the audio is coming from
media = luaviz.FfmpegCliBoostMedia.new(arg[1], SIZE)

-- layering & timing logic
viz = luaviz.Base.new(SIZE, media)
framerate = viz:get_framerate()

-- tells `viz` to only read FFT_SIZE audio samples from `media` per video frame (default 60fps)
viz:set_audio_frames_needed(FFT_SIZE)

-- iosevka is a great font, use hardware h264 encoder (available on my laptop)
if luaviz.os == 'linux' then
	font_path = '/usr/share/fonts/TTF/Iosevka-Regular.ttc'
	vcodec = 'h264_vaapi'
elseif luaviz.os == 'windows' then
	font_path = os.getenv('LocalAppData') .. '\\Microsoft\\Windows\\Fonts\\Iosevka-Regular.ttc'
	vcodec = 'h264_qsv'
end
viz:set_text_font(font_path)

-- set up song metadata object
font = luaviz.sfFont.new(font_path)
title_text = luaviz.sfText.new(font)
artist_text = luaviz.sfText.new(font)
smd = luaviz.SongMetadataDrawable.new(title_text, artist_text)

-- grab the album cover/art from the audio file
attached_pic = media:attached_pic()

if attached_pic then
	smd:set_album_cover(attached_pic, { 150, 150 })
end

title_text:setStyle(luaviz.sfTextStyle.Bold | luaviz.sfTextStyle.Italic)
title_text:setCharacterSize(24)
title_text:setFillColor({ 255, 255, 255, 150 })
artist_text:setStyle(luaviz.sfTextStyle.Italic)
artist_text:setCharacterSize(24)
artist_text:setFillColor({ 255, 255, 255, 150 })

smd:set_position({ 30, 30 })
smd:use_metadata(media)

---utility function from the c++ code
---@param margin integer
function set_spectrum_margin(margin)
	bss:set_rect({ { margin, margin }, { SIZE[1] - 2 * margin, SIZE[2] - 2 * margin } })
end

set_spectrum_margin(10)
-- make a "mirror" effect
bss:set_left_backwards(true)

if attached_pic then
	bg_spr = luaviz.Sprite.new(attached_pic)
	bg_spr:capture_centered_square_view()
	bg_spr:fill_screen(SIZE)

	bg_layer = viz:add_layer('bg', 0)
	bg_layer:add_drawable(bg_spr)
	-- we need to store the effects in variables for now, otherwise they get garbage collected...
	bg_blur = luaviz.Blur.new(7.5, 7.5, 15)
	bg_darken = luaviz.Mult.new(.75)
	bg_layer:add_effect(bg_blur)
	bg_layer:add_effect(bg_darken)
end

function do_fft_safely()
	-- make sure the StereoAnalyzer allocates enough space for the BarStereoSpectrum's bars
	bss:configure_analyzer(sa)
	-- using the configuration in FrequencyAnalyzer, perform fft for both channels and store the result in the StereoAnalyzer
	viz:perform_fft(fa, sa)
end

particles_layer = viz:add_layer('particles', 0)
particles_layer:add_drawable(cps)
particles_blur = luaviz.Blur.new(1, 1, 10)
particles_layer:add_effect(particles_blur)

particles_layer:set_orig_cb(function(_)
	do_fft_safely()
	if framerate == 60 then
		cps:update(sa)
	end
	-- handle other framerates, check ttviz::layers_init() for reference
end)

particles_layer:set_fx_cb(function(orig_rt, fx_rt, target)
	target:draw(fx_rt:sprite(), luaviz.sfBlendMode.Add)
	target:draw(orig_rt:sprite(), luaviz.sfBlendMode.Add)
end)

spectrum_layer = viz:add_layer('spectrum', 0)

-- register the object to be drawn by the layer
spectrum_layer:add_drawable(bss)

-- perform IMPORTANT logic before any drawables are drawn
-- (we don't need to draw onto the layer ourselves, hence the unused `_` parameter)
spectrum_layer:set_orig_cb(function(_)
	-- update the spectrum bar heights using the fft results
	bss:update(sa)
end)

-- blend the layer onto the final picture by adding colors (involves transparency)
spectrum_layer:set_fx_cb(function(orig_rt, fx_rt, target)
	target:draw(fx_rt:sprite(), luaviz.sfBlendMode.Add)
	target:draw(orig_rt:sprite())
end)

-- add subtle glow effect on the spectrum
spectrum_blur = luaviz.Blur.new(3, 3, 10)
spectrum_layer:add_effect(spectrum_blur)

-- a "final drawable" is something that depends on everything else being drawn first
viz:add_final_drawable(smd)

-- watch the viz!!!!!!!!!!!!!!!
viz:start_in_window(arg[0])
